### PR TITLE
Don't show context cancelled errors in attribute reader

### LINKED ISSUE #18997: Error on open last repository commit: "Unable to open checker for ..."
### Gitea Version

Gitea v1.16.3

### Git Version

Git Version: 2.34.1

### Operating System

Debian GNU/Linux 11 (bullseye) x86_64

### How are you running Gitea?

Gitea package from "_Gitea v1.16.3 release assets_": [gitea-1.16.3-linux-amd64.xz](https://github.com/go-gitea/gitea/releases/download/v1.16.3/gitea-1.16.3-linux-amd64.xz)

Gitea is running on external disk (sd card) mounted as local media via command line:
```
export GITEA_WORK_DIR=/media/user/sdportable/gitea/home
/media/user/sdportable/gitea/bin/gitea web -c /media/user/sdportable/gitea/home/app.ini
...

2022/03/04 10:43:34 cmd/web.go:102:runWeb() [I] Starting Gitea on PID: 174594
2022/03/04 10:43:34 cmd/web.go:150:runWeb() [I] Global init
2022/03/04 10:43:34 routers/init.go:106:GlobalInitInstalled() [I] Git Version: 2.34.1, Wire Protocol Version 2 Enabled
2022/03/04 10:43:34 routers/init.go:109:GlobalInitInstalled() [I] AppPath: /media/user/sdportable/gitea/bin/gitea
2022/03/04 10:43:34 routers/init.go:110:GlobalInitInstalled() [I] AppWorkPath: /media/user/sdportable/gitea/home
2022/03/04 10:43:34 routers/init.go:111:GlobalInitInstalled() [I] Custom path: /media/user/sdportable/gitea/home/custom
2022/03/04 10:43:34 routers/init.go:112:GlobalInitInstalled() [I] Log path: /media/user/sdportable/gitea/home/log
2022/03/04 10:43:34 routers/init.go:113:GlobalInitInstalled() [I] Configuration file: /media/user/sdportable/gitea/home/app.ini
2022/03/04 10:43:34 routers/init.go:114:GlobalInitInstalled() [I] Run Mode: Prod
2022/03/04 10:43:34 ...dules/setting/log.go:283:newLogService() [I] Gitea v1.16.3 built with GNU Make 4.1, go1.17.7 : bindata, sqlite, sqlite_unlock_notify
2022/03/04 10:43:34 ...dules/setting/log.go:330:newLogService() [I] Gitea Log Mode: Console(Console:)
2022/03/04 10:43:34 ...dules/setting/log.go:246:generateNamedLogger() [I] Router Log: Console(console:)
...
```

### Database

SQLite

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Description

Sometimes, not systematically, an error appears in the logs, when I click on the the last `commit id` in any repository, from all places: from 'user activity', from the specific repository 'home page', from the 'commits list', ...
When it happens the error is always the same:
```
2022/03/04 10:35:56 ...s/gitdiff/gitdiff.go:1418:func1() [E] Unable to open checker for 8fd941c7454eb37aafcac4a6d410d68669efdc5b. Error: failed to run attr-check. Error: context canceled
	Stderr: 
```
My 'report' was to know if this type of error can affect the integrity of the 'git repository'.

Note: as much as I could check the problem is only for the last commit of the repository!

I attach the complete log of a request that generated the error.

Thanks in advance.
Enrico

### Screenshots

A sample console log for the _error request_:

```
2022/03/04 10:35:56 Started GET /user/alpha/commit/8fd941c7454eb37aafcac4a6d410d68669efdc5b for [::1]:56774
2022/03/04 10:35:56 ...dules/git/command.go:143:RunWithContext() [D] /media/user/sdportable/gitea/home/data/gitea-repositories/user/alpha.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true rev-parse
2022/03/04 10:35:56 ...dules/git/command.go:143:RunWithContext() [D] /media/user/sdportable/gitea/home/data/gitea-repositories/user/alpha.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true show-ref --tags
2022/03/04 10:35:56 ...dules/git/command.go:143:RunWithContext() [D] /media/user/sdportable/gitea/home/data/gitea-repositories/user/alpha.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true cat-file --batch
2022/03/04 10:35:56 ...dules/git/command.go:143:RunWithContext() [D] /media/user/sdportable/gitea/home/data/gitea-repositories/user/alpha.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true cat-file --batch-check
2022/03/04 10:35:56 ...dules/git/command.go:143:RunWithContext() [D] /media/user/sdportable/gitea/home/data/gitea-repositories/user/alpha.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true show-ref --heads
2022/03/04 10:35:56 ...s/gitdiff/gitdiff.go:718:ParsePatch() [D] ParsePatch(1000, 5000, 100, ..., )
2022/03/04 10:35:56 ...s/charset/charset.go:135:DetectEncoding() [D] Detected encoding: utf-8 (fast)
2022/03/04 10:35:56 ...s/charset/charset.go:135:DetectEncoding() [D] Detected encoding: utf-8 (fast)
2022/03/04 10:35:56 ...dules/git/command.go:143:RunWithContext() [D] /media/user/sdportable/gitea/home/data/gitea-repositories/user/alpha.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true read-tree 8fd941c7454eb37aafcac4a6d410d68669efdc5b
2022/03/04 10:35:56 ...dules/git/command.go:143:RunWithContext() [D] /media/user/sdportable/gitea/home/data/gitea-repositories/user/alpha.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true check-attr --stdin -z --cached linguist-vendored linguist-generated linguist-language gitlab-language --
2022/03/04 10:35:56 ...dules/git/command.go:143:RunWithContext() [D] /media/user/sdportable/gitea/home/data/gitea-repositories/user/alpha.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true diff --shortstat 1ef546dfd6d69709e3a5a101415d9bce1c2c2001...8fd941c7454eb37aafcac4a6d410d68669efdc5b
2022/03/04 10:35:56 ...s/gitdiff/gitdiff.go:1418:func1() [E] Unable to open checker for 8fd941c7454eb37aafcac4a6d410d68669efdc5b. Error: failed to run attr-check. Error: context canceled
	Stderr: 
2022/03/04 10:35:56 ...dules/git/command.go:143:RunWithContext() [D] /media/user/sdportable/gitea/home/data/gitea-repositories/user/alpha.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true name-rev --exclude refs/tags/* --name-only --no-undefined 8fd941c7454eb37aafcac4a6d410d68669efdc5b
2022/03/04 10:35:56 ...dules/git/command.go:143:RunWithContext() [D] /media/user/sdportable/gitea/home/data/gitea-repositories/user/alpha.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c uploadpack.allowfilter=true describe --exact-match --tags --always 8fd941c7454eb37aafcac4a6d410d68669efdc5b
2022/03/04 10:35:56 ...s/context/context.go:198:HTML() [D] Template: repo/commit_page
2022/03/04 10:35:56 Completed GET /user/alpha/commit/8fd941c7454eb37aafcac4a6d410d68669efdc5b 200 OK in 30.093522ms
```

### PR DESCRIPTION
**Pull request by zeripath**
Fix #18997

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Review by Gusted — APPROVED** [review]

**Comment by lunny** [comment]
But we may need to log them?

File: modules/git/repo_attribute.go [hunk 1]
```diff
@@ -205,7 +205,7 @@ func (c *CheckAttributeReader) Run() error {
 			return nil
 		},
 	})
-	if err != nil && c.ctx.Err() != nil && err.Error() != "signal: killed" {
+	if err != nil && c.ctx.Err() != err && err.Error() != "signal: killed" {
```
Comment (wxiaoguang) [inline modules/git/repo_attribute.go]: It's difficult for a reader to understand what does `c.ctx.Err() != err` mean and why it should be ignored here ..... I think we need more comments for reasons and correctness.

**Comment by zeripath** [comment]
We don't need to log these.

The program being ended because the context has been cancelled is a normal thing and it should not be logged.

Similarly the "signal: killed" error will only occur due to the process being killed. This will happen due to context cancellation.

These are not errors and should not be logged.

**Comment by wxiaoguang** [comment]
Why `c.ctx.Err() is err` means `context is being cancelled`, but not using `Error() string matches "context canceled"`?

The other expression below is using `err.Error() matches "signal: killed"` to mean the process was killed .....

**Comment by zeripath** [comment]
What's your question here?

We can ignore both of these as they're expected errors.

**Comment by wxiaoguang** [comment]
> What's your question here?
> 
> We can ignore both of these as they're expected errors.

I was asking about the meaning about `c.ctx.Err() != err`. I still have difficulty to understand whether it is correct in all cases.

`c.ctx.Err() != err` is a very rare expression for this scenario, most developers just seldom see or use it.

### Question 1: Is it really correct?

For example:
* case 1: `c.ctx.Err() = "context canceled"` and `err = "other error"`?
* case 2: `c.ctx.Err() = "other error"` and `err = "context canceled"`?
* case 3: `c.ctx.Err() = "the-error"` and `err = "the-error"`?

Are they all correct for the expression `c.ctx.Err() != err`? Even it's correct, it's not obvious, it's hard to imagine that future maintainers can know why it works.

### Question 2: The error detection

The other expression is `err.Error() != "signal: killed"`, so could it be `err.Error() != "context canceled"` here? If not, what's the reason.

**Comment by wxiaoguang** [comment]
ps: to clarify .... the questions above are not a blocker. Maybe only I have the difficulty to understand the logic ....

If someone can confirm it's correct, just give the approval (and I would really appreciate if someone could help explain the logic and questions .... 🙏)

**Comment by zeripath** [comment]
Yes they are all correct. 

Read the docs for context err. Different errs can be returned from the context err

It doesn't matter what caused the context to be cancelled it is cancelled.

The signal killed error also needs to be ignored due to cmd.Wait sometimes returning that when the exec is killed by the context being out of date.

**Comment by wxiaoguang** [comment]
If it is the case, I think we should patch: `RunContext`

https://github.com/go-gitea/gitea/blob/98f54081d62d82213952ffe4bdbb459572da91b1/modules/git/command.go#L182-L186

```go
	if err := cmd.Wait(); err != nil && ctx.Err() != context.DeadlineExceeded { 
		return err 
	} 
	if ctx.Err() == context.Canceled {
		// if the command (context) is canceled by our code, we shouldn't treat such result as an error
		return nil
	}
	return ctx.Err()
```

**Review by wxiaoguang — APPROVED** [review]
IMO `err != context.Canceled` could be more clear and correct, and we should report `DeadlineExceeded` as error.

If you are sure that `c.ctx.Err() != err` is by purpose, the comment could be updated to `We should not pass up error due to the context being cancelled or it exceeds deadline`

File: modules/git/repo_attribute.go [hunk 2]
```diff
@@ -205,7 +205,9 @@ func (c *CheckAttributeReader) Run() error {
 			return nil
 		},
 	})
-	if err != nil && c.ctx.Err() != nil && err.Error() != "signal: killed" {
+	if err != nil && //                      If there is an error we need to return but:
+		c.ctx.Err() != err && //             1. We should not pass up error due to the context being cancelled
```
Comment (wxiaoguang) [inline modules/git/repo_attribute.go]:
```suggestion
		c.ctx.Err() != err && //             1. Ignore the context error if the context is cancelled or exceeds the deadline (RunWithContext could return c.ctx.Err() which is Canceled or DeadlineExceeded)
```

**Pull request merged**